builtin_array.go

```diff
@@ -1023,6 +1023,59 @@ func (r *Runtime) arrayproto_findIndex(call FunctionCall) Value {
 	return intToValue(-1)
 }
 
+func (r *Runtime) arrayproto_flat(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	l := toLength(o.self.getStr("length", nil))
+	depthNum := int64(1)
+	if len(call.Arguments) > 0 {
+		depthNum = call.Argument(0).ToInteger()
+	}
+	a := arraySpeciesCreate(o, 0)
+	r.flattenIntoArray(a, o, l, 0, depthNum, nil, nil)
+	return a
+}
+
+func (r *Runtime) flattenIntoArray(target, source *Object, sourceLen, start, depth int64, mapperFunction func(FunctionCall) Value, thisArg Value) int64 {
+	targetIndex, sourceIndex := start, int64(0)
+	for sourceIndex < sourceLen {
+		p := intToValue(sourceIndex)
+		if source.hasProperty(p.toString()) {
+			element := source.get(p, source)
+			if mapperFunction != nil {
+				element = mapperFunction(FunctionCall{
+					This:      thisArg,
+					Arguments: []Value{element, p, source},
+				})
+			}
+			if depth > 0 && isArray(element.ToObject(r)) {
+				elementLen := toLength(element.ToObject(r).self.getStr("length", nil))
+				targetIndex = r.flattenIntoArray(target, element.ToObject(r), elementLen, targetIndex, depth-1, nil, nil)
+			} else {
+				if targetIndex >= maxInt-1 {
+					panic(r.NewTypeError("Invalid array length"))
+				}
+				createDataPropertyOrThrow(target, intToValue(targetIndex), element)
+				targetIndex++
+			}
```

Comment:
`IsArray` should be applied directly to the element, so this block should be like this:
```suggestion
			var elementArray *Object
			if depth > 0 {
				if elementObj, ok := element.(*Object); ok && isArray(elementObj) {
					elementArray = elementObj
				}
			}
			if elementArray != nil {
				elementLen := toLength(elementArray.self.getStr("length", nil))
				targetIndex = r.flattenIntoArray(target, elementArray, elementLen, targetIndex, depth-1, nil, nil)
			} else {
				if targetIndex >= maxInt-1 {
					panic(r.NewTypeError("Invalid array length"))
				}
				createDataPropertyOrThrow(target, intToValue(targetIndex), element)
				targetIndex++
			}
```

Reply:
This did fix the tests with `undefined`
Thanks :)

Reply:
As for the rest, I quickly backported and checked them, they all passed for me. I believe it has something to do with babel...

This looks good enough for merging. If you are going to implement any optimisations they may go as a separate PR(s)